### PR TITLE
fix: remove redundant code in `export * from 'external'`    for esm format

### LINKED ISSUE #1940: [Bug]: esm `export * from 'externalized-dep'` still generating unnecessary code
### Reproduction link or steps

https://github.com/rolldown/repros/tree/esm-re-export

```js
// main.js
export * from "x"
```
```js
// rolldown.config.mjs
export default {
  input: './main.js',
  external: ['x']
}
```

### What is expected?

Output should just be

```js
export * from "x"
```

### What is actually happening?

Output includes tons of unnecessary code:

```js
import "x";

export * from "x"

//#region rolldown:runtime
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));

//#endregion
//#region main.js
var main_ns = {};
import * as import_x from "x";
__reExport(main_ns, import_x);

//#endregion
```

Most importantly, the line `import * as import_x from "x";` completely breaks treeshaking for any module that imports this one.

### System Info

```Shell
rolldown@nightly (0.12.2-snapshot-a2bb2ca-20240811003152)
```

### Any additional comments?

_No response_

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
### Description
Closed https://github.com/rolldown/rolldown/issues/1940

### CODE REVIEW

File: crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs [hunk 1]
```diff
@@ -112,25 +110,10 @@ impl<'me, 'ast> VisitMut<'ast> for ScopeHoistingFinalizer<'me, 'ast> {
               Module::External(importee) => {
                 match self.ctx.options.format {
                   rolldown_common::OutputFormat::Esm => {
-                    // Insert `import * as ns from 'ext'`
-                    // Insert `__reExport(exports, ns)`
-                    let re_export_fn_name = self.canonical_name_for_runtime("__reExport");
-                    let importer_namespace_name =
-                      self.canonical_name_for(self.ctx.module.namespace_object_ref);
                     let importee_namespace_name = self.canonical_name_for(rec.namespace_ref);
```
Comment (hyf0) [inline crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs]: You could get this in `EcmaModule#star_exports` , no need to collect them here.

**Comment by IWANABETHATGUY** [comment]
what the heck about the benchmark?

**Comment by IWANABETHATGUY** [comment]
> what the heck about the benchmark?

looks like just noise

File: crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs]: I tweak a little, but I still can't find a way to get `rec.namespace_ref` without get `rec`

File: crates/rolldown/src/stages/link_stage/mod.rs [hunk 2]
```diff
@@ -286,10 +284,12 @@ impl<'a> LinkStage<'a> {
...
-                      stmt_info.referenced_symbols.push(importer.namespace_object_ref.into());
-                      stmt_info
-                        .referenced_symbols
-                        .push(self.runtime.resolve_symbol("__reExport").into());
+                      if cjs_format {
+                        stmt_info.referenced_symbols.push(importer.namespace_object_ref.into());
+                        stmt_info
+                          .referenced_symbols
+                          .push(self.runtime.resolve_symbol("__reExport").into());
+                      }
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/stages/link_stage/mod.rs:292]: for `esm` we on-demand reference `importer.namespace_object_ref`

**Pull request merged**